Bluetooth: controller: Towards support for CIS

wopu-ot opened the pull request:
Work towards support for connected isochronous channels in the controller.

wopu-ot committed:
Bluetooth: controller: Define ULL/LLL interface for CIS central and periph

Define interface between ULL and LLL for CIS central and peripheral.

Signed-off-by: Wolfgang Puffitsch <wopu@demant.com>

mtpr-ot reviewed (commented):
Looks very good. Not 100% sure if all LLL fields are there, or some of them may not be needed, but it's an excellent starting point.

asmk-ot commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -241,6 +243,12 @@ struct node_rx_ftr {
+struct node_rx_iso_meta {
+	uint64_t payload_number; /* 39-bit cisPayloadNumber */
+	uint32_t timestamp;      /* Time of reception */
+	uint8_t  status;         /* Status of reception (OK/not OK) */
```
Where will this be set?

asmk-ot reviewed (approved):
Looks like a good starting point

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_conn_iso.h:
```diff
@@ -0,0 +1,68 @@
+	struct lll_conn_iso_strm_rxtx tx; /* TX parameters */
+
+	/* Event and payload counters */
+	uint64_t event_count;       /* 39-bit cisEventCount */
+	uint64_t rx_payload_number; /* 39-bit cisPayloadNumber */
```
Would it make sense to define these as 

```
	uint64_t event_count : 39;       /* 39-bit cisEventCount */
	uint64_t rx_payload_number : 39;  /* 39-bit cisPayloadNumber */
```
So that we cannot assign more than 39 bits to them (assuming that the compiler will catch value-out-of-bound errors).

Thalley reviewed (approved):
Generally LGTM; added a small comment which we can apply if we want, but is by no measure necessary.

Thalley commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -241,6 +243,12 @@ struct node_rx_ftr {
 
+/* Meta-information for isochronous PDUs in node_rx_hdr */
+struct node_rx_iso_meta {
+	uint64_t payload_number; /* 39-bit cisPayloadNumber */
```
```suggestion
	uint64_t payload_number :39  ; /* cisPayloadNumber */
```

wopu-ot replied:
Sorry for overlooking that one, thanks for pointing it out.

wopu-ot committed:
Bluetooth: controller: Add init/reset stubs for CIS central/peripheral ULL

Add stub functions to initialize and reset CIS central and CIS
peripheral roles in ULL.

Signed-off-by: Wolfgang Puffitsch <wopu@demant.com>

wopu-ot committed:
Bluetooth: controller: CIS HCI event generation

Implement generation of CIS Request and CIS Established events in HCI
layer.

Signed-off-by: Wolfgang Puffitsch <wopu@demant.com>

wopu-ot committed:
Bluetooth: controller: Add CIS-related LL control PDU types

Add definitions for LL_CIS_REQ, LL_CIS_RSP, and LL_CIS_IND PDUs.

Signed-off-by: Wolfgang Puffitsch <wopu@demant.com>

mtpr-ot commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -418,6 +418,10 @@ enum pdu_data_llctrl_type {
 	PDU_DATA_LLCTRL_TYPE_PHY_RSP = 0x17,
 	PDU_DATA_LLCTRL_TYPE_PHY_UPD_IND = 0x18,
 	PDU_DATA_LLCTRL_TYPE_MIN_USED_CHAN_IND = 0x19,
+	PDU_DATA_LLCTRL_TYPE_CIS_REQ = 0x1F,
```
You have to fill in the missing ones for the LUT to work:
PDU_DATA_LLCTRL_TYPE_CTE_REQ = 0x1A
PDU_DATA_LLCTRL_TYPE_CTE_RSP = 0x1B
PDU_DATA_LLCTRL_TYPE_PERIODIC_SYNC_IND = 0x1C
PDU_DATA_LLCTRL_TYPE_CLOCK_ACCURACY_REQ = 0x1D
PDU_DATA_LLCTRL_TYPE_CLOCK_ACCURACY_RSP = 0x1E

wopu-ot committed:
Bluetooth: controller: Refactor LLCP PDU length checks

Refactor LLCP PDU length checks to avoid constructing a look-up-table.

Signed-off-by: Wolfgang Puffitsch <wopu@demant.com>

wopu-ot committed:
Bluetooth: controller: Add support for ISO feature bits

Add support for ISO feature bits. This includes extending fields that
hold features to 64 bits to be able to handle the host-controlled
"Isochronous Channels" feature.

Signed-off-by: Wolfgang Puffitsch <wopu@demant.com>

Thalley commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -554,6 +554,22 @@ config BT_CTLR_PERIPHERAL_ISO
 	bool "LE Connected Isochronous Stream Peripheral [EXPERIMENTAL]" if BT_LL_SW_SPLIT
 	select BT_CTLR_SET_HOST_FEATURE
 
+config BT_CTLR_CONN_ISO_GROUPS
+	int "LE Connected Isochronous Groups"
+	depends on BT_CTLR_CENTRAL_ISO || BT_CTLR_PERIPHERAL_ISO
+	range 1 240
+	default 2
+	help
+	  Maximum supported CIGs.
+
+config BT_CTLR_CONN_ISO_STREAMS
+	int "LE Connected Isochronous Streams per Group"
+	depends on BT_CTLR_CENTRAL_ISO || BT_CTLR_PERIPHERAL_ISO
+	range 1 31
+	default 1
+	help
+	  Maximum supported CISes per CIG.
```
I think this is fine for now. I do wonder if a collective pool of streams would be a better solution (that's what we have implemented in the host). 

Imagine a use case where you need typically need to do 2 CIG with 2 CiS each, but once in a while you need to 1 CIG with 5 CIS and occasionally 1 BIG with 5 BIS, but never at the same time. That would be a total of 10 CIS and 5 BIS streams you need to preallocate, even if you never use more than 5 streams (CIS or BIS) at a time. 

Of course you are probably more aware of the potential usecases than I am, but I think there might be some optimization to be found by using a collective pool of streams, possibly even using the same streams for CIS/BIS.

As mentioned this is fine for now, but it's food for thought :)

wopu-ot replied:
I think a pool of CISes would be a good idea as an optimization. It will not necessarily make this setting go away though. If we have a pool of 4 CISes and but every CIG can contain at most 2 CISes, we only want 2 pointers to CISes in a CIG.

wopu-ot replied:
On second thought, let's get this right from the beginning and have separate pools for CIGs and CISes.

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_conn_iso.h:
```diff
@@ -0,0 +1,70 @@
+	struct lll_conn *conn;
+
+	/* Connection parameters */
+	uint8_t  access_addr[4];    /* Access address */
```
Just out of curiosity: Why is a this a 4 octet array instead of an uint32_t?

wopu-ot replied:
Because then it looks the same on little-endian and big-endian machines (and the endianness is not easily changed by accident), which helps a lot when comparing big-endian/little-endian traces.

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_conn_iso.h:
```diff
@@ -0,0 +1,70 @@
+	struct lll_hdr hdr;
+
+	uint8_t  num_cis;  /* Number of CISes in this CIG */
+#if defined(CONFIG_BT_CTLR_CONN_ISO_STREAMS)
+	struct lll_conn_iso_strm strms[CONFIG_BT_CTLR_CONN_ISO_STREAMS];
+#endif /* CONFIG_BT_CTLR_CONN_ISO_STREAMS */
```
Do we need the guard here? 

`CONFIG_BT_CTLR_CONN_ISO_STREAMS` doesn't seem to be guarded by anything in the Kconfig, so it should always be visible/defined.

wopu-ot replied:
`CONFIG_BT_CTLR_CONN_ISO_STREAMS` has a `depends on BT_CTLR_CENTRAL_ISO || BT_CTLR_PERIPHERAL_ISO`, so some guard is needed.

Thalley replied:
Oh wow, I missed that somehow :s

Thalley commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -570,6 +574,69 @@ struct pdu_data_llctrl_min_used_chans_ind {
+#else
+#error "Unsupported endianness"
+#endif
+	uint8_t  c_sdu_interval[3];
+	uint8_t  p_sdu_interval[3];
+	uint16_t c_max_pdu;
```
For this and a couple of places below, you could do a `uint64_t` bitfield when you have 24 + 24 + 16 => 64 bits. If you prefer 3 byte arrays, that's just fine as well :)

wopu-ot replied:
3-byte arrays seem to be the thing to use for 24-bit values in other places, so I'll keep them.

Thalley commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -570,6 +574,69 @@ struct pdu_data_llctrl_min_used_chans_ind {
+	uint8_t  p_sdu_interval[3];
+	uint16_t c_max_pdu;
+	uint16_t p_max_pdu;
+	uint8_t  nse;
+	uint8_t  sub_interval[3];
```
Similarly this could also be a `uint32_t` bitfield

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn_iso_internal.h:
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) 2020 Demant
```
```suggestion
 * Copyright (c) 2021 Demant
```

wopu-ot replied:
Done.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_peripheral_iso_internal.h:
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) 2020 Demant
```
```suggestion
 * Copyright (c) 2021 Demant
```

wopu-ot replied:
Done.

Thalley reviewed (commented):
A couple of comments; none that really needs to be fixed, but are potential fixes.

asmk-ot commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -241,6 +243,12 @@ struct node_rx_ftr {
+/* Meta-information for isochronous PDUs in node_rx_hdr */
+struct node_rx_iso_meta {
+	uint64_t payload_number : 39; /* cisPayloadNumber */
+	uint32_t timestamp;           /* Time of reception */
```
There is probably an opportunity here to align these definitions with the ones doen in iso adaptation layer, once this PR is in the bag I can look at following suit. Should the stream be recorded also?

wopu-ot replied:
I would expect that the stream can be identified through the handle, but I agree that this struct will change depending on how the ISO AL evolves.

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3135,11 +3151,52 @@ static void le_cis_established(struct pdu_data *pdu_data,
 			       struct net_buf *buf)
 {
+	struct bt_hci_evt_le_cis_established *sep;
+	struct lll_conn_iso_strm_rxtx *lll_cis_c;
+	struct lll_conn_iso_strm_rxtx *lll_cis_p;
+	struct lll_conn_iso_strm *lll_cis;
+	struct node_rx_conn_iso_est *est;
+	struct ll_conn_iso_group *cig;
+	struct ll_conn_iso_strm *cis;
```
```suggestion
	struct lll_conn_iso_stream_rxtx *lll_cis_c;
	struct lll_conn_iso_stream_rxtx *lll_cis_p;
	struct node_rx_conn_iso_estab *est;
	struct lll_conn_iso_stream *lll_cis;
	struct ll_conn_iso_stream *cis;
	struct ll_conn_iso_group *cig;
```

wopu-ot replied:
Done

cvinayak reviewed (changes requested):
Quickly reviewed for implementation style, accept a minor naming, all looks good to me. Have not reviewed for functional correctness nor tested it. If anything needs fixing that I discover when using this code, can be taken up then.
